caffe2/perfkernels/sve_emblookup_codegen.py

```diff
@@ -0,0 +1,469 @@
+                "            svAll,\n"
+                "            vwgt,\n"
+                "            svcvt_f32_f16_x(\n"
+                "                svAll,\n"
+                "                svreinterpret_f16_u32(svld1uh_u32(\n"
+                "                    svAll, reinterpret_cast<const uint16_t*>("
+                f"&ip[{regid} * vLen])))),\n"  # noqa
+                f"            vsum{regid});"
+            )
+        elif InType == "at::BFloat16":
+            code.append(
+                f"        vsum{regid} = svmad_f32_x(\n"
+                "            svAll,\n"
+                "            vwgt,\n"
+                "            svreinterpret_f32_u32(svlsl_n_u32_x(\n"
+                "                svAll,\n"
+                "                svld1uh_u32(\n"
+                "                    svAll, reinterpret_cast<const uint16_t*>("
+                f"&ip[{regid} * vLen])),\n"
+                "                16)),\n"  # noqa
+                f"            vsum{regid});"
+            )
+        elif InType == "uint8_t":
+            code.append(
+                f"        vsum{regid} = svmad_f32_x(\n"
+                "            svAll,\n"
+                "            vwgt,\n"
+                "            svcvt_f32_u32_x(svAll,"
+                f" svld1ub_u32(svAll, &ip[{regid} * vLen])),\n"  # noqa
+                f"            svadd_f32_x(svAll, vsum{regid}, vbio));"
+            )
+        else:
+            assert False
+
+        return code
+
+    code = []
+    code.append("    // unrolling " + str(uf) + " times")
+
+    code.append(
+        "    for ("
+        + IndexType
+        + " i = 0; i < output_size; ++i) {"
+    )
+
+    code.append("      " + OutType + "* const op = &out[i * block_size];")
+    if use_offsets:
+        code.append(
+            "      if (pos != offsets[i] - offsets[0]) {\n"
+            + "        return false;\n"
+            + "      }"
+        )
+    else:
+        exit(1)
+    # Initialise vector sum registers
+    for i in range(0, uf):
+        code.append("      svfloat32_t vsum" + str(i) + " = svdup_n_f32(0);")
+
+    # inner loop
+    code.append("""\
+      int64_t start_offset = offsets[i];
+      int64_t end_offset = offsets[i + 1];""")
+    code.append(
+        "      for ("
+        + "int64_t"
+        + " j = start_offset; j < end_offset; ++j) {"  # noqa
+    )
+
+    code.append("        const auto idx = indices[pos];")
+    code.append(
+        "        if (idx < 0 || idx >= data_size) {\n"
+        + "          return false;\n"
+        + "        }"
+    )
+
+    if InType == "uint8_t":
+        code.append("        " + OutType + " wgt = 1.f;")
+        code.append("        " + OutType + " bio{};")
+        code.append("        if (weights) {")
+        code.append(
+            "          wgt = weights[IS_WEIGHT_POSITIONAL ? (j - start_offset) : pos];"  # noqa
+        )
+        code.append("        }")
+
+        if fused:
```

Comment:
Since we already validate this in the main function, it would be better practice to do something like `assert(not fused, "fused is not supperted` at the top of this function

Reply:
Code path removed